core/src/services/github_contents/core.rs

```diff
@@ -0,0 +1,289 @@
+        self.client.send(req).await
+    }
+
+    pub fn sign(&self, req: request::Builder) -> Result<request::Builder> {
+        let req = req.header(header::USER_AGENT, constants::USER_AGENT);
+
+        Ok(req.header(
+            header::AUTHORIZATION,
+            format_authorization_by_bearer(&self.token)?,
+        ))
+    }
+}
+
+impl GithubContentsCore {
+    pub async fn get_file_sha(&self, path: &str) -> Result<String> {
+        let resp = self.stat(path).await?;
+
+        match resp.status() {
+            StatusCode::OK => {
+                let headers = resp.headers();
+
+                let sha = parse_etag(headers)?;
+
+                let Some(sha) = sha else {
+                    return Err(Error::new(
+                        ErrorKind::Unexpected,
+                        "No ETag found in response headers",
+                    ));
+                };
+
+                Ok(sha.trim_matches('"').to_string())
+            }
+            StatusCode::NOT_FOUND => Err(Error::new(ErrorKind::NotFound, "File not found")),
+            _ => Err(parse_error(resp).await?),
+        }
+    }
+
+    pub async fn stat(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let path = build_abs_path(&self.root, path);
+
+        let url = format!(
+            "https://api.github.com/repos/{}/{}/contents/{}",
+            self.owner,
+            self.repo,
+            percent_encode_path(&path)
+        );
+
+        let req = Request::head(url);
+
+        let req = self.sign(req)?;
+
+        let req = req
+            .header("Accept", "application/vnd.github.raw+json")
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.send(req).await
+    }
+
+    pub async fn get(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let path = build_abs_path(&self.root, path);
+
+        let url = format!(
+            "https://api.github.com/repos/{}/{}/contents/{}",
+            self.owner,
+            self.repo,
+            percent_encode_path(&path)
+        );
+
+        let req = Request::get(url);
+
+        let req = self.sign(req)?;
+
+        let req = req
+            .header("Accept", "application/vnd.github.raw+json")
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.send(req).await
+    }
+
+    pub async fn upload(&self, path: &str, bs: Bytes) -> Result<Response<IncomingAsyncBody>> {
+        let sha = self.get_file_sha(path).await;
+
+        let path = build_abs_path(&self.root, path);
+
+        let url = format!(
+            "https://api.github.com/repos/{}/{}/contents/{}",
+            self.owner,
+            self.repo,
+            percent_encode_path(&path)
+        );
+
+        let req = Request::put(url);
+
+        let req = self.sign(req)?;
+
+        let mut req_body = json!({
+            "message": "opendal upload",
```

Comment:
The message is `commit message` for this operation, right? We should allow user to custom it in the future. But currently, we can add path and current time into the message.

Reply:
Can you give me an example?

Reply:
How about `Write {path} at {time} via opendal`?

Reply:
Why we need {path}? it already know where writed.

Reply:
> Why we need {path}? it already know where writed.

Make this commit more meaningful from the user side. For example, an opendal user visits the commits page and finds a lot of messages only have time.